Tests/PactConsumerSwiftTests/AnimalServiceClient.swift

```diff
@@ -1,164 +1,268 @@
+        failure(self.prepareError(error))
       }
     }
+
   }
 
+  ///
+  /// Get a specific alligator by `id`
+  ///
+  /// Calls either succes that takes an Array<Animal>, or NSError on failure as a closure
+  ///
   open func getAlligator(_ id: Int, success: @escaping (Animal) -> Void, failure: @escaping (NSError?) -> Void) {
-    self.performRequest("\(baseUrl)/alligators/\(id)", decoder: decodeAnimal) { animal, nsError in
-      if let animal = animal {
-        success(animal)
-      } else {
-        if let error = nsError {
-          failure(error)
-        } else {
-          failure(NSError(domain: "", code: 42, userInfo: nil))
-        }
+    request(
+      urlString: "\(baseUrl)/alligators/\(id)"
+    ) { data, response, error in
+      guard error == nil else { failure(error as NSError?); return }
+      guard let data = data else { failure(self.prepareError(error)); return }
+
+      do {
+        let alligator = try JSONDecoder().decode(Animal.self, from: data)
+        success(alligator)
+      } catch {
+        failure(self.prepareError(error))
       }
     }
   }
 
+  ///
+  /// Get a list of Animals that live in an environment
+  ///
+  /// Calls closure with Array<Animal> as parameter
+  ///
+  /// - parameter live: A string representing the environment where an Animal lives (eg: "water")
+  ///
   open func findAnimals(live: String, response: @escaping ([Animal]) -> Void) {
-    let liveEncoded = live.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed)!
-    self.performRequest("\(baseUrl)/animals?live=\(liveEncoded)", decoder: decodeAnimals) { animals, nsError in
-      if let animals = animals {
+    request(
+      urlString: "\(baseUrl)/animals",
+      queryParameters: ["live": live]
+    ) { data, requestResponse, error in
+      guard let data = data else {
+        response([]);
+        return
+      }
+
+      do {
+        let debugging = String(decoding: data, as: UTF8.self)
+        debugPrint(debugging)
+        let animals = try JSONDecoder().decode(Array<Animal>.self, from: data)
         response(animals)
+      } catch {
+        response([])
       }
     }
+
   }
 
+  ///
+  /// Define what animal an alligator eats
+  ///
+  /// - parameter animal: A String representing an animal as food to an alligator (eg: "Pigeon")
+  ///
   open func eat(animal: String, success: @escaping () -> Void, error: @escaping (Int) -> Void) {
-    self.performRequest("\(baseUrl)/alligator/eat", method: "patch", parameters: ["type": animal], decoder: decodeString) { string, nsError in
-      if let localErr = nsError {
-        error(localErr.code)
+    request(
+      urlString: "\(baseUrl)/alligator/eat",
+      method: .patch,
+      bodyParameters: ["type": animal]
+    ) { data, response, responseError in
+      if let response = response as? HTTPURLResponse {
+        (200..<400).contains(response.statusCode) ? success() : error(response.statusCode)
       } else {
-        success()
+        error(500)
```

Comment:
Shouldn't this be returning the status code from the service?

Reply:
Is/Was a _bad_ client implementation for the test purpose so I personally wasn't too fussed at what it does